man/man1/wclock.1
.TH WCLOCK 1 "Version 1, Release 4" "W Window System" "W PROGRAMS"
.SH NAME
wclock \- analog clock for W
.SH SYNOPSIS
wclock [-g \fIgeometry\fP] [-b \fIpattern id\fP] [-r] [-h]
.SH DESCRIPTION
\fIWclock\fP is an analog clock.  It's window is updated about once a
minute or when a key is pressed.  You can get rid of \fIwclock\fP with
the 'q' key or by sending it a signal.
.SH OPTIONS
You may use the `-g' option to dictate size and position.  With color
servers you can use the '-c' option to set clock background to one of
the shared server colors (normally 0-7).  For larger clock sizes you
could set a background pattern with '-b' option (value as hexadecimal).
\'-r' makes \fIwclock\fP to draw into root window and `-h' just displays
the usage line.
.PP
IMHO \fIwclock\fP looks best at very small (<80x80) or large sizes
(>240x240). For example '-g 31,31 -c 3' or '-g 480,480 -b 1511'.
.SH SEE ALSO
.IR w_setpattern (3)
.SH ORIGINAL AUTHOR
TeSche, itschere@techfak.uni-bielefeld.de
.SH LAST CHANGED
2/98 by Eero Tamminen


wt/CHANGES
++oddie 9/2000:
- fixed listbox/filesel resizing & added minimum size setopt.

++bsittler 7/2000:
- Fixed vt color/reverse and attribute resetting
- Add HTTP/1.1 'Host' header to Wetscape requests
- Other minor changes (netgames, wvt, waudio)

++jcarlson 7/2000:
- WTINYSCREEN changes

++eero, 5/1999:
- Fixed key handling for special keys in Wt/getstring.c & wedit/wedit.c.

++eero, 1/1999:
- Added support for new W special keys so that now on W platforms which
implement special key mappings, you don't need to use Emacs shortcuts
with text widgets (or remap your console cursor keys)...


------ notes for version released with W R1PL0 ----------

ADDITIONAL WIDGETS

- Arrow, a simple button with an arrow drawn into it.

- Dial, a dial widget where you can position a needle into desired
  angle.  This isn't in the W toolkit library itself as then you would
  need to link math library to all of your Wt programs.  You'll find it
  in the warzone game tar-ball.

- Edittext, mainly intended for simple one page stuff, but as this
  spaghetti can do some additional stuff like paragraph formatting,
  there's a real abuse built on this, the WEdit.  As edittext widget is
  large and buggy, it's in the WEdit directory instead of the toolkit.

- Menu, a strip of buttons from which popups pop-up.  Both the popups
  and menus use a structure defined in Wt.h instead of strings that Kay
  was using.  The reason is that I want callbacks to follow
  automatically hierarchically opened popups.

- Popup, hierarchical popups.  Can include subpopups, images, lines and
  in addition items can also be disabled and checked.

- Range, a scrollbar composite with which you can select a value between
  given limits.

- Select, a getstring widget coupled with an arrow openable listbox from
  which items can be selected into a getstring widget.


OTHER CHANGES

- vt can now do everything wterm  can do + user can rebind the colors
  in wtrc.

- Added '.wtconfig' file and made makefiles use the W compilation
  configuration file '.config' in parent directory.

- wt_do_event() doesn't call wt_dispatch_event() anymore with
  EVENT_GADGET/GADGET_EXIT event as those might not have valid window
  pointer.  If you want to check for abnormal (unboud) exit, check the
  wt_return value.  GADGET_EXIT returns a negative number and
  GADGET_CLOSE (in shell widget) a positive number now.

- Added xterm mouse events, reverse mode, fd/bgcolor setting + improved
  double/triple click timing in VT widget.

- Aliases a couple new tags into others in HTML widget and added
  HTML header (ignored) and TITLE parsing.

- Widget event processing functions return now unrecognized or
  non-processed WEVENTs or NULL instead of long.  Assumption is that if
  function returns WEVENT, the widget whose action it was, is still
  around.  If in doubt, use NULL.  Unrecognized key events should be
  returned for focus change though.  (actually it could still have been
  an int, but returning a WEVENT offers a possibility for event mapping
  widgets.)

- Added wt_global structure containing Wt configuration variables and
  made widgets to use it.  wt_init() will now read these variables from
  ~/.wtrc + a list of (variable, value) pairs.  Widget specific
  initialization code in their init() functions may read these with the
  wt_variable() function.

- HTML widget fonts can be configured by above method.

- Added wt_loadfont() function which will use the user specified Wt
  default if widget doesn't specify some font parameter.  It also exits
  program if Wt default fonts can't be loaded, because otherwise
  unavailable fonts may chrash naive programs.

- Changed widgets to reflect the new w_loadfont() syntax (NULL/zero
  values tell Wlib/server to use server defaults).

- Added new WT_FONTSIZE define.  The earlier WT_FONT setopt sets now the
  font *family* name, new one sets the font size.

- Getopting WT_FONT returns WFONT* pointer also for above reason.

- Removed 'fontname' members from widget structures.  Needed information
  and more can be gotten from the 'font' member if needed.

- Fileselector has now separate lists for directory and file names.
  Both of course have an autolocator (type & return in addition to point
  & click).

- Wt_keys.h defines values for some special keys like cursors etc.  Half
  of the defines are Emacs style and half what the special keys itself
  give or are in use on other platforms.

- Changed Wt option defines.  There were a couple overlapping ones.


------- Kay's original notes: ----------------------------------------------

- added some missing close() and truncate the file to zero length before
  writing a new selection in selection.c.

- better selection handling in VT widget.

- made everything resizeable.

- applied patches by eero and fixed some more bugs myself.

- implemented a simple clipboard and changed getstring and vt widget to use it.

- fixed several serious bugs in the html widget.

- wrote VT widget.

- added an option to the icon editor to sync edit field with display
  field (simon).

- added missing special chars to html widget (simon).

- added code to the fileselector to set a default file name.

- lots of changes by eero (button.c, fileselect.c, listbox.c, getstring.c), some
  fixes to his changes by me.

- added missing wt_ungetfocus() to listbox and getstring.

- wrote Tk-like packer widget.

- new fileselect widget and testprogram by eero. Added better filename
  matching (eeros version was a joke, really).

- getstring and listbox updates by eero.

- changed the scrollbar widget's user interface to become more user friendly.
  See scrollbar.doc for details on how to use it. Also changed viewport and
  listbox to use the new interface.

- added listbox_focus(). Otherwise listbox cannot receive keyboard focus...

- big bugfix: when wt_delete() is called on a widget all pending timers must
  be canceled! (affected list, iconedit and getstring).


apps/wclock.c
/*
 * wclock.c, a part of the W Window System
 *
 * Copyright (C) 1994-1996 by Torsten Scherer and
 * Copyright (C) 1995-1998 by Eero Tamminen
 *
 * This package is free software; you can redistribute it and/or modify it
 * under the terms specified in the docs/COPYRIGHTS file coming with this
 * package.
 *
 * -- an analog clock
 *
 * CHANGES
 * 12/94, TeSche:
 * - Changed for W1R0
 * 11/95, Eero:
 * - Uses a lookup table instead of sin & cos
 * - calculations in fixed point instead of floats
 * - Clock face scales to the window size
 * - Shows only hours and minutes
 * 96, TeSche:
 * - adapted for W1R2
 * 11/97, Eero:
 * - resizable
 * 12/97, jps:
 * - adapted to the swallow feature of wlaunch
 * - added long options
 * 2/98, Eero:
 * - clock hand widths and middle 'stone' size according to clock size.
 * - root and bg pattern options. You can't have 'transparent' clock
 *   with root option because we have to use M_DRAW mode.
 * 6/98, Eero:
 * - background color can be changed.
 */

#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <time.h>
#include <Wlib.h>
#include <getopt.h>


static struct option long_options [] = {
  { "bg", required_argument, NULL, 'b' },
  { "color", required_argument, NULL, 'c' },
  { "geometry", required_argument, NULL, 'g' },
  { "parent", required_argument, NULL, 'p' },
  { "root", no_argument, NULL, 'r' },
  { "help", no_argument, NULL, 'h' },
  { NULL, 0, NULL, 0}
};

/* window minimum / default size */
#define MIN_SIZE	30
#define DEF_SIZE	80

#define PROPERTIES	(W_MOVE | W_RESIZE | EV_KEYS)


typedef struct {
  short	hxe, hye;
  short	mxe, mye;
} POINTERS;


static WWIN *window;
static short midx, midy;
static short maxxr, maxyr;
static short dot;


/* offset and draw clock pointers */
static void draw(POINTERS *p)
{
  w_pbox(window, midx - (dot>>1), midy - (dot>>1), dot+1, dot+1);
  w_line(window, midx - (p->hxe >> 2), midy - (p->hye >> 2),
         midx + p->hxe, midy + p->hye);
  w_line(window, midx - (p->mxe >> 3), midy - (p->mye >> 3),
         midx + p->mxe, midy + p->mye);
}

/* offset and clear the clock pointers */
static void clear(POINTERS *p)
{
  w_dline(window, midx - (p->hxe >> 2), midy - (p->hye >> 2),
          midx + p->hxe, midy + p->hye);
  w_dline(window, midx - (p->mxe >> 3), midy - (p->mye >> 3),
          midx + p->mxe, midy + p->mye);
}

/* calculate clock pointer offsets from the middle */
static void offset(short *x, short *y, short sixty)
{
  short sini[16] =			/* sin(2*PI / 60 * sixty) * (1<<14) */
  {
        0,  1713,  3406,  5063,  6664,  8192,  9630, 10963,
    12176, 13255, 14189, 14968, 15582, 16026, 16294, 16384
  };
  short xx, yy, index = sixty % 15;

  switch(sixty / 15)
  {
    case 3:
      xx = -sini[15-index];
      yy = -sini[index];
      break;
    case 2:
      xx = -sini[index];
      yy = sini[15-index];
      break;
    case 1:
      xx = sini[15-index];
      yy = sini[index];
      break;
    default:
      xx = sini[index];
      yy = -sini[15-index];
  }
  *x = ((long)xx * maxxr) >> 14;
  *y = ((long)yy * maxyr) >> 14;
}

/* update clock pointers */
static void update(struct tm *loc)
{
  static POINTERS p = {0, 0, 0, 0};

  /* clear old pointers */
  clear(&p);

  /* offset from the middle */
  offset(&p.hxe, &p.hye, (loc->tm_hour % 12) * 5 + (loc->tm_min + 6) / 12);
  offset(&p.mxe, &p.mye, loc->tm_min);

  /* shorten hour pointer */
  p.hxe = p.hxe * 2 / 3;
  p.hye = p.hye * 2 / 3;

  /* draw new pointers */
  draw(&p);
}

/* set globals and draw the clock 'numbers' */
static void draw_clock(short width, short height, short color, ushort pattern)
{
/* half_size / x offsets */
#define P_DIST   6
#define N_DIST  12	
  short i, xp, yp, size, skip;

  midx = width / 2;
  midy = height / 2;
  maxxr = midx - midx / N_DIST;
  maxyr = midy - midy / N_DIST;

  size = MIN(midx, midy) / 40;

  if (size > 1) {
    w_setlinewidth(window, (size<<1) + 1);
    skip = 1;
  } else {
    w_setlinewidth(window, 1);
    pattern = MAX_GRAYSCALES;	/* white bg */
    skip = 5;
  }

  if (color > 0) {
    w_setBackgroundColor(window, color);
  }
  w_setpattern(window, pattern);
  w_dpbox(window, 0, 0, width, height);

  for (i = 0; i < 60; i += skip)
  {
    offset(&xp, &yp, i);
    xp += midx;
    yp += midy;
    if(size)
    {
      if (i % 5) {
        /* minutes between... */
	w_pbox(window, xp - (size>>1), yp - (size>>1), size, size);
      } else {
        /* ...hours */
	w_pbox(window, xp - size, yp - size, (size<<1), (size<<1));
      }
    }
    else
      w_plot(window, xp, yp);
  }
  maxxr -= midx / P_DIST;
  maxyr -= midy / P_DIST;
  /* getting this comfortable with linewidth at all clock sizes really took
   * some experimenting...
   */
  dot = (size*size>>1) + (size<<1) + 2;
}



static void check_size(short *width, short *height)
{
  /* limit size */
  if (*width < MIN_SIZE) {
    if (*width <= 0)
      *width = DEF_SIZE;
    else
      *width = MIN_SIZE;
  }
  if (*height < MIN_SIZE) {
    if(*height <= 0)
      *height = DEF_SIZE;
    else
      *height = MIN_SIZE;
  }
}

/* open window */
static int open_clock (char *name, short width, short height,
                       short xp, short yp, WWIN *parent)
{
  if (parent) {
    if (!(window = w_createChild (parent, width, height, W_NOBORDER))) {
      fprintf(stderr, "%s: can't create child window\n", name);
      return -1;
    }
  } else {
    if (!(window = w_create(width, height, PROPERTIES))) {
      fprintf(stderr, "%s: can't create window\n", name);
      return -1;
    }
    limit2screen(window, &xp, &yp);
  }

  if (w_open(window, xp, yp) < 0)
  {
    fprintf(stderr, "%s: can't open window\n", name);
    w_exit();
    return -1;
  }

  return 0;
}

static void terminate()
{
  w_exit();
  exit(0);
}



static void usage (char *name) {
  fprintf (stderr, "\
Usage: %s [OPTION]...\n\
The program that lets you know what time it is.\n\
\n\
  -b, --bg         background pattern id (hex)\n\
  -c, --color      background color index\n\
  -g, --geometry   set size and position\n\
  -r, --root       draw clock to the root window\n\
  -h, --help       display this help and exit\n\
\n\
", name);
}


int main(int argc, char *argv[])
{
  short	width = 0;
  short height = 0;
  short xp = UNDEF;
  short yp = UNDEF;
  short color = 0;			/* default bg color */
  ushort pattern = MAX_GRAYSCALES;		/* white bg */
  short use_root = 0, show_usage = 0;
  ulong parentid = 0;
  WWIN *parent = NULL;
  WEVENT *ev;
  struct tm *loc;
  long t;
  int c;

  while ((c = getopt_long (argc, argv, "b:c:g:p:rh", long_options, NULL)) != -1) {
    switch (c) {

      case 'g':
        scan_geometry (optarg, &width, &height, &xp, &yp);
        break;

      case 'b':
        sscanf (optarg, "%hx", &pattern);
        break;

      case 'p':
        sscanf (optarg, "%lx", &parentid);
        break;

      case 'c':
        color = atoi(optarg);
	break;

      case 'r':
        use_root = 1;
	break;

      default:
        show_usage = 1;
    }
  }

  if (show_usage) {
    usage (argv[0]);
    exit (0);
  }

  if (!w_init()) {
    fprintf (stderr, "%s: can't connect to wserver\n", argv[0]);
    exit (1);
  }

  if (use_root) {
    window = WROOT;
    width = window->width;
    height = window->height;
  } else {

    if (parentid) {
      xp = yp = 0;
      parent = w_winFromID(parentid);
      height = parent->height;
      width = parent->width;
    } else {
      check_size(&width, &height);
    }

    if (open_clock (argv[0], width, height, xp, yp, parent)) {
      return -1;
    }
  }

  w_setmode(window, M_DRAW);
  draw_clock(width, height, color, pattern);

  signal(SIGTTOU, SIG_IGN);

  /* set exit handlers (one time only, so won't need 'reliable' ones) */
  signal(SIGHUP,  terminate);
  signal(SIGINT,  terminate);
  signal(SIGQUIT, terminate);
  signal(SIGABRT, terminate);
  signal(SIGTERM, terminate);

  while(42)
  {
    t = time(0);
    loc = localtime(&t);
    update(loc);

    /* watch out for 'q' / `exit' event while idling a min */
    ev = w_queryevent(NULL, NULL, NULL, 60000);
    if(!ev)
      continue;

    switch(ev->type)
    {
      case EVENT_GADGET:
        if (ev->key == GADGET_EXIT)
	  terminate();
	break;

      case EVENT_KEY:
        if (ev->key == 'q')
	  terminate();
	break;

      case EVENT_RESIZE:
	width = ev->w;
	height = ev->h;
	check_size(&width, &height);
	w_move (ev->win, ev->x, ev->y);
	w_resize (ev->win, width, height);

	/* updates global size variables too */
	draw_clock(width, height, color, pattern);
	break;
    }
  }
  return 0;
}
